horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java

```diff
@@ -982,4 +980,150 @@ public QueryResult queryDataSet(Integer datasetId, String jsonpath, boolean arra
       }
       return result;
    }
+
+   private void transform(Run run, String role) {
+      if (run == null || role == null) {
+         log.errorf("Transformation parameters error: run %s, role %s", run, role);;
+      }
+      Collection<RunService.RunSchema> runSchemas = getSchemas(run, role);
+      Test test = Test.findById(run.testid);
+      List<JsonNode> nodesToProcess = new ArrayList<JsonNode>();
+      boolean matched = false;
+
+      if (!runSchemas.isEmpty()){
+         for ( RunSchema s : runSchemas ){
+            if (hasMatch(s, run.data)) {
+               nodesToProcess.add(run.data);
+            }
+            int nodeSize = nodesToProcess.size();
+
+            if (nodeSize == 0) {
+               new DataSet(run.start, run.stop, "",
+                     test.id, run.data, run, 0, run.owner, run.access)
+                     .persist();
+            } else {
+               Collection<Transformer> transformers = test.transformers;
+               if (transformers.isEmpty()) {
+                  new DataSet(run.start, run.stop, "",
+                        test.id, run.data, run, 0, run.owner, run.access)
+                        .persist();
+               } else {
+                  for (Transformer t : transformers){
+                     StringBuilder array = new StringBuilder("[");
+                     boolean atStart = true;
+                     for (int nodePos = 0; nodePos <= nodeSize; nodePos += 1 ) {
+                        if (!atStart) {
+                           array.append(",");
+                           atStart = false;
+                        }
+                        JsonNode n = nodesToProcess.get(nodePos);
+                        if (!t.extractors.isEmpty()) {
+                           for ( NamedJsonPath e : t.extractors ){
+                              QueryResult result = queryData(run.id, e.jsonpath, s.uri, true);
+                              String value = result.value;
+                              if (null != value && !"[]".equals(value)) {
+                                 JsonNode nodes = Util.toJsonNode(result.value);
+                                 for (JsonNode node : nodes){
+                                    if (!atStart) {
+                                       array.append(",");
+                                    }
+                                    array.append(node.asText());
+                                 }
+                                 matched = true;
+                              }
+                           }
+                        }
+                     }
+                     array.append("]");
+
+                     DataSet ds = null;
+                     try {
+                        if (t.function == null || t.function.isBlank()) {
+                           ds = new DataSet(run.start, run.stop
```

Comment:
I think you misunderstood the final part when transformer results are merged: it is *NOT* that each Transformer should produce it's own DataSet. There should be one DataSet for each object in the array created by the Transformer, and if some transformers produce and array and others just single object, the single-objects are duplicated to match the longest array produced by any Transformer.

Reply:
I'll present a more complex example with input
```
{
  "foos": {
    "$schema": "foo",
    "foo": "bar",
    "aaa": 42
  },
  "bars": {
    "$schema": "bar",
    "xyz": [1, 2, 3]
  }
}
```
Schema "foo", Transformer A:
```
foo: $.foo
aaa: $.aaa
({foo, aaa}) => {
  return ({
    "$schema": "nested-foos",
    "nested": { foo, aaa }
  })
}
```
Schema "bar", Transformer B
```
xyz: $.xyz
({xyz}) => {
  return xyz.map(x => ({
     "$schema": "individual-bars",
     "value": x
  }))
}
```

Reply:
Transformer A will return
```
{
  "$schema": "nested-foos",
  "nested": {
    "foo": "bar",
    "aaa": 42
  }
}
```
Transformer B will return
```
[
  {
    "$schema": "individual-bars",
    "value": 1
  },
  {
    "$schema": "individual-bars",
    "value": 2
  }, … // third item omitted for brevity
]
```
The result of these will be 3 DataSets and each will look like
```
[
  {
    "$schema": "nested-foos",
    "nested": {
      "foo": "bar",
      "aaa": 42
    }
  },
  {
    "$schema": "individual-bars",
    "value": 1 // the other datasets will contain values 2 and 3 respectively
  },
]
```

Reply:
there must be a condition here. Only 1 Transformer can have a output producing an array result. In this case B. All other Transformers must produce a scalar result. Correct ?

Reply:
Is it safe to assume that Transformer A and B executors are executed on both schemas "foo" and "bar" ? The outcome is expected to be empty for some of these.

Reply:
there must be a condition here. The condition being only **one** Transformer may produce a result with multiple elements. How would the results for DataSet be padded out for 3 Transformers, producing 1, 3 and 8 elements ?

Reply:
Not really; if you have multiple transformers that create array result, all of them *should* produce the array of the same length. In that case the datasets would have elements with respective indexes into those arrays.
I say *should* because we have to deal with the case that they don't do that - the options I can think of are:

* somehow declare and error and don't produce anything
* ignore missing elements, e.g. T1: `[ A, B, C]`, T2: `[ X, Y ]` -> DS1: `[ A, X ]`, DS2: `[ B, Y ]`, DS3: `[ C ]`
* repeat elements from beginning: DS3: `[C, X]`

I find the third option very confusing, my choice would be the second one

Reply:
I was inclined to think something between the first and second choice.
 The first array wins. Any that follow that don't match the size are reported as an error.
 It's easy to explain the logic to users.

Reply:
By first you mean shortest? Or just "random" as the order in which you fetch the listeners is not defined?

With this approach you could signal that something's wrong only on the Run. That makes it a bit harder to spot. With longest array those 'incomplete' datasets could visually indicate the missing data.

Reply:
ok.  incomplete datasets it is.